Keeper: persistent watches are not removed on session close (mntr shows non-zero zk_watch_count)

### Describe what's wrong

Keeper doesn’t clear persistent watches on session close. After closing a client with PERSISTENT and PERSISTENT_RECURSIVE watches, mntr still shows 3 watches. This violates ZooKeeper semantics, leaks server state, skews metrics, and degrades performance.

Feature was added in https://github.com/ClickHouse/ClickHouse/pull/88813

### Does it reproduce on the most recent release?

Yes

### How to reproduce

- Start Keeper
- Create node /testPersistentCleanup
- Add watches:
  - AddWatch(PERSISTENT) on /testPersistentCleanup
  - AddWatch(PERSISTENT_RECURSIVE) on /testPersistentCleanup
- Close the client session (stop + close)
- Check mntr

Example - failures were observed in https://github.com/ClickHouse/ClickHouse/pull/92377/files (https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=92377&sha=e421db872935f24c0e59c6209359d8ff2e3c1304&name_0=PR):
```
client.add_watch(NODE_PATH, cb, AddWatchMode.PERSISTENT)
client.add_watch(NODE_PATH, cb, AddWatchMode.PERSISTENT_RECURSIVE)
destroy_zk_client(client)
data = keeper_utils.send_4lw_cmd(cluster, node1, cmd="mntr")
assert "zk_watch_count\t0" in data  # Fails: returns 3
```

### Expected behavior

- All watches (including persistent, persistent-list, and persistent-recursive) for the closed session are removed.
- mntr: zk_watch_count is 0 after session close.

### Error message and/or stacktrace

mntr shows 3 watches remaining post-close:
```
File: test_keeper_persistent_watches/test.py:317 - in test_persistent_watches_cleanup_on_close
    assert "zk_watch_count\t0" in data
E   AssertionError: assert 'zk_watch_count\t0' in 'zk_version\tv25.12.1.597-testing-e421db872935f24c0e59c6209359d8ff2e3c1304\nzk_avg_latency\t0\nzk_max_latency\t3\nzk_min_latency\t0\nzk_packets_received\t7\nzk_packets_sent\t5\nzk_num_alive_connections\t2\nzk_outstanding_requests\t0\nzk_server_state\tstandalone\nzk_znode_count\t8\n**zk_watch_count\t3**\nzk_ephemerals_count\t0\nzk_approximate_data_size\t1412\nzk_key_arena_size\t0\nzk_latest_snapshot_size\t264\nzk_open_file_descriptor_count\t72\nzk_max_file_descriptor_count\t524288\nzk_followers\t0\nzk_synced_followers\t0\n'
```

### Additional context

Why `3` is correct but wrong
- Keeper registers persistent watches in multiple maps:
  - PERSISTENT → persistent_watches + persistent_list_watches (2 entries)
  - PERSISTENT_RECURSIVE → persistent_recursive_watches (1 entry)
- Total per the reproduction: 2 + 1 = 3
- These entries remain after session close → **leak**

**Code pointers:**
https://github.com/ClickHouse/ClickHouse/blob/2dd0b0dc1a9024a4867174859af993409a1e91c1/src/Coordination/KeeperStorage.cpp#L4020 clearDeadWatches skips persistent types (advances iterator, doesn’t erase) for:
- PERSISTENT_WATCH
- PERSISTENT_LIST_WATCH
- PERSISTENT_RECURSIVE_WATCH

**ZooKeeper semantics**: persistent watches persist across events, not across session termination. They are session-scoped and must be removed on close/expire. Example Reference (of closing watches on session close) - https://github.com/JDRaftKeeper/RaftKeeper/issues/60#issuecomment-1514596319

**Impact**: Incorrect metrics/diagnostics, Memory growth

--- a/src/Coordination/KeeperStorage.cpp
+++ b/src/Coordination/KeeperStorage.cpp
@@ -4018,11 +4018,13 @@ void KeeperStorageBase::clearDeadWatches(int64_t session_id)
                 erase_session_from_map(watches, watch_path);
                 break;
             case WatchType::PERSISTENT_WATCH:
-                [[fallthrough]];
+                erase_session_from_map(persistent_watches, watch_path);
+                break;
             case WatchType::PERSISTENT_LIST_WATCH:
-                [[fallthrough]];
+                erase_session_from_map(persistent_list_watches, watch_path);
+                break;
             case WatchType::PERSISTENT_RECURSIVE_WATCH:
-                ++watch_it;
+                erase_session_from_map(persistent_recursive_watches, watch_path);
                 break;
         }
     }

--- a/tests/integration/test_keeper_persistent_watches/test.py
+++ b/tests/integration/test_keeper_persistent_watches/test.py
@@ -328,3 +328,19 @@ def test_persistent_watches_cleanup_on_close(started_cluster):
             break
         time.sleep(0.1)
     assert "zk_watch_count\t0" in data
+
+def test_clear_watches(started_cluster):
+    node1.restart_clickhouse()
+    keeper_utils.wait_until_connected(cluster, node1)
+
+    client = get_fake_zk(node1)
+    NODE_PATH = "/testPersistentCleanup"
+
+    def cb(_):
+        pass
+
+    client.add_watch(NODE_PATH, cb, AddWatchMode.PERSISTENT)
+    client.add_watch(NODE_PATH, cb, AddWatchMode.PERSISTENT_RECURSIVE)
+    destroy_zk_client(client)
+    data = keeper_utils.send_4lw_cmd(cluster, node1, cmd="mntr")
+    assert "zk_watch_count\t0" in data  # Fails: returns 3

--- a/tests/integration/test_keeper_persistent_watches/test.py
+++ b/tests/integration/test_keeper_persistent_watches/test.py
@@ -343,4 +343,4 @@ def test_clear_watches(started_cluster):
     client.add_watch(NODE_PATH, cb, AddWatchMode.PERSISTENT_RECURSIVE)
     destroy_zk_client(client)
     data = keeper_utils.send_4lw_cmd(cluster, node1, cmd="mntr")
-    assert "zk_watch_count\t0" in data  # Fails: returns 3
+    assert "zk_watch_count\t0" in data

